core/src/main/java/com/linecorp/armeria/common/util/DefaultLimitedBlockingTaskExecutor.java

```diff
@@ -0,0 +1,240 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.collect.ImmutableList;
+
+final class DefaultLimitedBlockingTaskExecutor implements LimitedBlockingTaskExecutor {
+    private final BlockingTaskExecutor delegate;
+
+    private final SettableIntSupplier limitSupplier;
+
+    private final AtomicInteger taskCounter = new AtomicInteger(0);
+
+    DefaultLimitedBlockingTaskExecutor(BlockingTaskExecutor delegate, SettableIntSupplier limitSupplier) {
+        this.delegate = delegate;
+        this.limitSupplier = limitSupplier;
+    }
+
+    @Override
+    public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+        return delegate.schedule(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                command.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, delay, unit);
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
+        return delegate.schedule(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                return callable.call();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, delay, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period,
+                                                  TimeUnit unit) {
+        return delegate.scheduleAtFixedRate(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                command.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, initialDelay, period, unit);
+    }
+
+    @Override
+    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable command, long initialDelay, long delay,
+                                                     TimeUnit unit) {
+        return delegate.scheduleAtFixedRate(() -> {
+            try {
+                taskCounter.incrementAndGet();
+                command.run();
+            } finally {
+                taskCounter.decrementAndGet();
+            }
+        }, initialDelay, delay, unit);
+    }
+
+    @Override
+    public void shutdown() {
+        delegate.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+        return delegate.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return false;
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return delegate.isShutdown();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return delegate.awaitTermination(timeout, unit);
+    }
+
+    @Override
+    public <T> Future<T> submit(Callable<T> task) {
+        return delegate.submit(() -> {
+            try {
+                taskCounter.incrementAndGet();
```

Comment:
Shouldn't the counter be increased as soon as the task is submitted rather than when it's consumed? i.e.
```java
boolean submitted = false;
taskCounter.incrementAndGet();
try {
  future = delegate.submit(() -> {
    try {
      return task.call();
    } finally {
      taskCounter.decrementAndGet();
    }
  });
  submitted = true;
  return future;
} finally {
  if (!submitted) {
    taskCounter.decrementAndGet();
  }
}    
```